app/controllers/tasks_controller.rb

```diff
@@ -0,0 +1,103 @@
+class TasksController < ApplicationController
+  def index
+    @tasks = Task.all
+  end
+
+  def show
+    @task = Task.find_by(id: params[:id])
+
+    if @task.nil?
+      redirect_to tasks_path
+      return
+    end
+
+    if @task.completed
+      @status = "completed"
+    elsif @task.due_date && @task.due_date < DateTime.now
+      @status = "overdue"
+    else
+      @status = "in-progress"
+    end
+
+    if @task[:completed]
+      @change = "Incomplete"
+    else
+      @change = "Complete"
+    end
+  end
+
+  def new
+    @task = Task.new
+  end
+
+  def create
+    task = Task.new(task_params)
+
+    is_successful = task.save
+
+    if is_successful
+      redirect_to task_path(task.id)
+    else
+      head :not_found # refactor this?
+    end
+  end
+
+  def edit
+    @task = Task.find_by(id: params[:id])
+    if @task.nil?
+      redirect_to tasks_path
+      return
+    end
+  end
+
+  def update
+    task = Task.find_by(id: params[:id])
+
+    if task.nil?
+      redirect_to tasks_path
+      return
+    end
+
+    is_successful = task.update(task_params)
+
+    if is_successful
+      redirect_to tasks_path
+    else
+      head :not_found
+    end
+  end
+
+  def destroy
+    task = Task.find_by(id: params[:id])
+
+    if task.nil?
+      head :not_found # refactor
+    else
+      task.destroy
+      redirect_to tasks_path
+    end
+  end
+
+  def mark_complete
+    task = Task.find_by(id: params[:id])
+
+    if task.completed
+      cd = nil
+    else
+      cd = DateTime.now
+    end
+
+    if task.nil?
+      head :not_found # refactor ?
+    else
+      task.update(completed: !(task[:completed]), completion_date: cd)
+      redirect_to tasks_path
+    end
```

Comment:
This method works just fine! It's definitely functional and correct.

However, it's not quite what I would call "elegant." There are a couple of things that could be refactored. You check `task.completed` first, and _then_ check if `task` is `nil`. Does that seem like the most logical order? Also, this separates the logic of setting completion date from updating it, but it doesn't feel... like those pieces of logic should be separated.

This is absolutely just a _code style_ comment, and not one on functionality, but I felt like it might be interesting to think about.